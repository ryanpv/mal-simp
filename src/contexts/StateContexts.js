import React, { useContext, useState, } from 'react'


const StateContext = React.createContext();

export function useStateContext() {
  return useContext(StateContext)
};

export function StateProvider({ children }) {
  const [loading, setLoading] = useState(false);
  const [show, setShow] = useState(false);
  const [animeDetails, setAnimeDetails] = useState({}); // needed additional state because MAL API does not allow certain field groups
  const [currentPage, setCurrentPage] = useState(1);
  const [animeList, setAnimeList] = useState({});
  const [currentTrailer, setCurrentTrailer] = useState({})
  const [searchResults, setSearchResults] = useState({})
  const [offset, setOffset] = React.useState(0)
  const [categoryList, setCategoryList] = React.useState({}) // state for list of users' categories *** WHY NOT []??? ***
  const [lastAddedCategory, setLastAddedCategory] = React.useState("");
  const [additionalContent, setAdditionalContent] = React.useState([]);
  const [errorMessage, setErrorMessage] = React.useState('')
  const [malUserDetails, setMalUserDetails] = React.useState({})
  const [malLoginMessage, setMalLoginMessage] = React.useState('No MAL user currently logged in.')





  const value = {
    show,
    setShow,
    offset,
    loading,
    setLoading,
    errorMessage,
    setErrorMessage,
    malLoginMessage,
    setMalLoginMessage,
    malUserDetails,
    setMalUserDetails,
    setOffset,
    animeDetails,
    setAnimeDetails,
    currentPage,
    setCurrentPage,
    animeList,
    setAnimeList,
    currentTrailer,
    setCurrentTrailer,
    searchResults,
    setSearchResults,
    categoryList,
    setCategoryList,
    lastAddedCategory,
    setLastAddedCategory,
    additionalContent,
    setAdditionalContent

  }

  return (
    <StateContext.Provider value={value}>
      { !loading && children }
    </StateContext.Provider>
  )
}
