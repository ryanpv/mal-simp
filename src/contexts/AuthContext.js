import React, { useContext, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { GoogleAuthProvider, signInWithPopup, createUserWithEmailAndPassword, signInWithEmailAndPassword, onAuthStateChanged, 
    signOut, sendPasswordResetEmail } from "firebase/auth";
import { auth } from '../Accounts/firebase';
import { useStateContext } from './StateContexts';

const AuthContext = React.createContext();
const providerGoogle = new GoogleAuthProvider();

export function useAuth() {
  return useContext(AuthContext);
};

export function AuthProvider({ children }) {
  const serverUrl = process.env.NODE_ENV === 'production' ? process.env.REACT_APP_DEPLOYED_SERVER : process.env.REACT_APP_SERVER_BASEURL
  const [error, setError] = useState('');
  const [currentUser, setCurrentUser] = useState();
  const navigate = useNavigate();
  const { setCategoryList, setErrorMessage, loading, setLoading } = useStateContext();


  function signup(email, password) {
    createUserWithEmailAndPassword(auth, email, password)
      .then((userCredential) => {
        const user = userCredential.user;
        console.log(user);
      })
      .catch((error) => {
        const errorCode = error.code
        const errorMessage = error.message
        console.log(errorCode);
        setError(errorMessage)
      });
  };

   function login(email, password) {
    signInWithEmailAndPassword(auth, email, password)
      .then(async (userCredential) => userCredential.user)
      .catch((error) => {
        const errorCode = error.code
        const errorMessage = error.message
        console.log(errorCode);
        setError(errorMessage);
      });
  };

  function logout() {
    setCurrentUser("");
    setCategoryList({});
    malLogout();
    navigate('/');
    return signOut(auth);
  };

  function resetPassword(email) {
    sendPasswordResetEmail(auth, email);
  };

  async function malLogout() {
    try {
      await fetch(`${ serverUrl }/clear-mal-cookie`, { credentials: 'include'})
      window.location.reload();
    } catch (err) {
      console.log(err);
    };
  };


  function loginWithGoogle(){
    signInWithPopup(auth, providerGoogle)
      .then(async (result) => {
        const credential = GoogleAuthProvider.credentialFromResult(result);
        const token = credential.accessToken;
        const user = result.user;

      })
      .catch((error) => {
        const errorCode = error.code;
        const errorMessage = error.message;
        const email = error.customData.email; // email of user
        const credential = GoogleAuthProvider.credentialFromError(error);
      });
      navigate('/');
  };

  React.useEffect(() => {
    const unsubscribe = onAuthStateChanged(auth, async (user) => {
      if (user) {
        await fetch(`${ serverUrl }/login-session`, {
          method: 'POST',
          credentials: 'include',
          headers: {
            "Content-type": "application/json"
          },
          body: JSON.stringify({ accessToken: user.accessToken, isRegUser: user.isRegUser })
        });
        setCurrentUser(user.displayName ? user.displayName : user.email);
        setLoading(false);
        // console.log('firebase token acquired');
        // console.log(user.accessToken && 'firebase token acquired');
      } else {
        setErrorMessage('No user detected. Please log in or refresh the page.');
        console.log('no user logged in');
      };
    });
    return unsubscribe;
  }, []);

  const value = {
    signup,
    login,
    loginWithGoogle,
    logout,
    resetPassword,
    setLoading,
    error,
    setError,
    currentUser,
    setCurrentUser,
  };


  return (
    <AuthContext.Provider value={ value }>
      { !loading && children }
    </AuthContext.Provider>
  );
};
