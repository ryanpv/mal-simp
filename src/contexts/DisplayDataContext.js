import React, { useContext } from 'react'
import { useStateContext } from './StateContexts';

const DisplayContext = React.createContext();

export function useDisplayContext() {
  return useContext(DisplayContext)
};

export function DisplayDataProvider({ children }) {
  const baseUrl = process.env.NODE_ENV === 'development' && process.env.REACT_APP_SERVER_BASEURL
  const { setShow, setAnimeDetails, setCurrentPage } = useStateContext();

    const handleShow = async (value) => {
      setShow(true)
      setCurrentPage(1)
      const getAnimeContent = await fetch(`${ baseUrl }/anime/${value.id}/videos,synopsis,mean,num_episodes`)
      const contentResults = await getAnimeContent.json();
      setAnimeDetails(contentResults)
      // console.log(contentResults.videos.length > 0 ? contentResults.videos[0].url : 'n/a');
      // console.log(contentResults);
    }


  const value = {
    handleShow
  }


  return (
    <DisplayContext.Provider value={ value }>
      { children }
    </DisplayContext.Provider>
  )
}
