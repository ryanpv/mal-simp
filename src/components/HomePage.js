import React from 'react'
import { Button, Row, Col, Card, Container } from 'react-bootstrap';
import { useStateContext } from '../contexts/StateContexts';
import { useDisplayContext } from '../contexts/DisplayDataContext';
import { useAuth } from '../contexts/AuthContext';
import { async } from '@firebase/util';

export default function HomePage() {
  const clientId = process.env.REACT_APP_MAL_CLIENT_ID
  const { animeList, setAnimeList, errorMessage, setErrorMessage, malUserDetails } = useStateContext();
  const { handleShow } = useDisplayContext();
  const [offset, setOffset] = React.useState(0)
  const { currentUser, setLoading } = useAuth();
  const baseUrl = process.env.NODE_ENV === 'development' && process.env.REACT_APP_SERVER_BASEURL



  async function fetchServer() {
    // const code_challenge = await getCode();
    try {
      const getCode = await fetch(`${ baseUrl }/create-challenge`, { headers: { 'Content-Type': 'application/json' }, credentials:"include" })
      const getChallenge = await getCode.json();

      await window.open(`https://myanimelist.net/v1/oauth2/authorize?response_type=code&client_id=${ clientId }&code_challenge=${ getChallenge }&redirect_uri=http://localhost:3000/logcallback`, "_self")
      
    } catch (error) {
      console.log(error);
    }
  }

  console.log(process.env.NODE_ENV === 'development');

  React.useEffect(() => {
    async function fetchRecommendedAnime() {
      try {
        if (malUserDetails.id) {

          const fetchRecommended = await fetch(`${ baseUrl }/user-recommendations/${ offset }`, { credentials: 'include' })
          const recommendationResults = await fetchRecommended.json();
          setAnimeList(recommendationResults);
          
          
          console.log('recommendations fetched', recommendationResults);
        }
         return ; 
      } catch (err) {
        if (err) {
          setErrorMessage('Log in to MAL to see recommendations')
        }
        // console.log(err);
      }
    }
    fetchRecommendedAnime();

  }, [offset, malUserDetails])


  async function incrementOffset(e) {
    e.preventDefault();
    setOffset(prevOffset => prevOffset + 5);
  }

  function decrementOffset(e) {
    e.preventDefault();
    setOffset(prevOffset => prevOffset - 5);
  }

  async function malLogout() {
    try {
      await fetch(`${ baseUrl }/clear-mal-cookie`, { credentials: 'include'})
      window.location.reload();
    } catch (err) {
      console.log(err);
    }
  }


  return (
    <>
    <div className='w-100 text-center mt-1 mb-4'>
      <h2>Welcome to MAL-Simp. A web-app that simplifies MyAnimeList</h2>
      </div>

    <div className='w-100 text-center'>
      { malUserDetails.name ? 
        <Button onClick={ () => malLogout() }>Log out of MAL</Button> 
        : <Button onClick={ () => fetchServer() }>Log In to MyAnimeList.net</Button> 
        }
    </div>

    { malUserDetails.name ? 
      <div className='text-center'>
        <h2>Anime recommendations for <strong><i>{ malUserDetails.name }</i></strong> :</h2>
      </div>
    : <h2>{ errorMessage }</h2>
    }
    
    <Container className='fluid'>
      <Row xs={1} md={5} className="g-4">
        { animeList.data ? animeList.data.map(recs => { return (
          <Col key={ recs.node.id } >
            <Card onClick={() => handleShow({ id: recs.node.id }) } bg="light" style={ { height: '100%', cursor: "pointer" } }>
              <Card.Img variant='top' src={ recs.node.main_picture.medium } />
              <Card.Body>
                <strong as="h6">{ recs.node.title }</strong>
              </Card.Body>
              <Card.Footer>Score: { recs.node.mean }</Card.Footer>
            </Card>
          </Col>
          );
        })
        : null
        }
      </Row>
      { animeList.paging ?
        <div className='w-100 text-center mt-2 mb-2'>
          { animeList.paging.previous ? <Button size='sm' onClick={(e) => decrementOffset(e)}>Previous</Button> : null }
          &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
          { animeList.paging.next ? <Button size='sm' onClick={(e) => incrementOffset(e)}>Next</Button> : null }
        </div> 
        : null
      }
      </Container>

    </>
  )
}

// create button that sends this code_challenge to server. server endpoint should fetch the API URL with code challenge as a parameter (step 2).
// make sure redirect URI goes back to server endpoint because client secret is required for token retrieval 