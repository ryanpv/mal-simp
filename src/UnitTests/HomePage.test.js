import React from "react";
import '@testing-library/jest-dom/extend-expect';
import { render, cleanup, } from '@testing-library/react';
import { StateProvider } from "../contexts/StateContexts";
import { AuthProvider } from "../contexts/AuthContext";
import { DisplayDataProvider } from "../contexts/DisplayDataContext";
import { BrowserRouter } from "react-router-dom";

import HomePage from "../components/HomePage";

  afterEach(cleanup)

  it('should take a snapshot', async () => {
    const { asFragment } = render(
      <BrowserRouter>
        <StateProvider>
          <AuthProvider>
            <DisplayDataProvider>
              <HomePage />
            </DisplayDataProvider>
          </AuthProvider>
        </StateProvider>
      </BrowserRouter>
    )


    // expect(await screen.findByText(/loaded/i).toBeInTheDocument())   
    expect(asFragment(<HomePage />)).toMatchSnapshot()

  })

  