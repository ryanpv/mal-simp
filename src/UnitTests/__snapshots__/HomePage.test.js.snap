// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="w-100 text-center mt-1 mb-4"
  >
    <h2>
      Welcome to MAL-Simp. A web-app that simplifies MyAnimeList
    </h2>
  </div>
  <div
    class="w-100 text-center"
  >
    <button
      class="btn btn-primary"
      type="button"
    >
      Log In to MyAnimeList.net
    </button>
  </div>
  <h2 />
  <div
    class="fluid container"
  >
    <div
      class="g-4 row row-cols-md-5 row-cols-1"
    />
  </div>
</DocumentFragment>
`;
